### PR TITLE
fix double-counting and non-deterministic results in `summarysize`

### LINKED ISSUE #53061: Base.summarysize returns different memory usage values on different runs
MWE:

```julia
mutable struct B
    x::Union{Float64, Tuple{Float64, Float64}}
    y::Union{Float64, Tuple{Float64, Float64}}
end

ex = B[B(rand(), (rand(),rand())) for _ in 1:10^5];
```

now if you execute:

```julia
julia> Base.summarysize(ex)
7776776

julia> Base.summarysize(ex)
7820648

julia> Base.summarysize(ex)
7923000

julia> Base.summarysize(ex)
7911824
```

This can become very extreme in some situation, I saw a 2x difference between runs on a more complex vector of structs.

Tested in 1.10 and nightly, same behaviour

### PR DESCRIPTION
**Pull request by JeffBezanson**
fixes #53061

### CODE REVIEW

**Comment by Tortar** [comment]
I think this fix should be backported to 1.10/1.11, right?

File: src/datatype.c [hunk 1]
```diff
@@ -2253,6 +2253,39 @@ JL_DLLEXPORT size_t jl_get_field_offset(jl_datatype_t *ty, int field)
...
+    const uint16_t *ptrs16 = (const uint16_t *)jl_dt_layout_ptrs(ly);
+    const uint32_t *ptrs32 = (const uint32_t*)jl_dt_layout_ptrs(ly);
+    uint32_t fld;
+    if (ly->flags.fielddesc_type == 0)
+        fld = ptrs8[i];
+    else if (ly->flags.fielddesc_type == 1)
+        fld = ptrs16[i];
+    else
+        fld = ptrs32[i];
+    return ((jl_value_t**)v)[fld];
```
Comment (vtjnash) [inline src/datatype.c:2273]: This should be an explicit relaxed load

**Comment by IanButterworth** [comment]
This introduced a failure on `test i686-linux-gnu`
```
  | 2024-06-21 13:47:24 EDT | misc                                             (10) \|         failed at 2024-06-21T17:47:24.853
  | 2024-06-21 13:47:24 EDT | Test Failed at /cache/build/tester-amdci5-10/julialang/julia-buildkite/julia-01af3e921f/share/julia/test/misc.jl:590
  | 2024-06-21 13:47:25 EDT | Expression: abs(summarysize(z) - 640000) / 640000 <= 0.01
  | 2024-06-21 13:47:25 EDT | Evaluated: 0.18746875 <= 0.01
  | 2024-06-21 13:47:25 EDT
```

**Comment by vtjnash** [comment]
We could revert, but I don't know if that may cause problems for the backport PR

**Comment by IanButterworth** [comment]
`@test_broken` on linux i686 and open an issue? I can do that if that sounds good

**Pull request merged**